sdk/examples/xo_python/sawtooth_xo/processor/handler.py

```diff
@@ -47,164 +47,235 @@ def namespaces(self):
+
+        # 5. Log for tutorial usage
+        if action == "create":
+            _display("Player {} created a game.".format(signer[:6]))
+
+        elif action == "take":
+            _display(
+                "Player {} takes space: {}\n\n".format(signer[:6], space) +
+                _game_data_to_str(
+                    upd_board, upd_state, upd_player1, upd_player2, name))
+
+        # 6. Put the game data back in state storage
+        _store_state_data(
+            state_store, game_list, name,
+            upd_board, upd_state,
+            upd_player1, upd_player2)
+
+
+def _unpack_transaction(transaction):
+    header = TransactionHeader()
+    header.ParseFromString(transaction.header)
+
+    # The transaction signer is the player
+    signer = header.signer_pubkey
+
+    try:
+        # The payload is csv utf-8 encoded string
+        name, action, space = transaction.payload.decode().split(",")
+    except ValueError:
+        raise InvalidTransaction("Invalid payload serialization")
 
+    _validate_transaction(name, action, space)
+
+    if action == 'take':
+        space = int(space)
+
+    return name, action, space, signer
+
+
+def _validate_transaction(name, action, space):
+    if not name:
+        raise InvalidTransaction('Name is required')
+
+    if '|' in name:
+        raise InvalidTransaction('Name cannot contain "|"')
+
+    if not action:
+        raise InvalidTransaction('Action is required')
+
+    if action not in ('create', 'take'):
+        raise InvalidTransaction('Invalid action: {}'.format(action))
+
+    if action == 'take':
         try:
-            # The payload is csv utf-8 encoded string
-            name, action, space = transaction.payload.decode().split(",")
+            assert int(space) in range(1, 10)
+        except (ValueError, AssertionError):
+            raise InvalidTransaction('Space must be an integer from 1 to 9')
+
+
+def _validate_game_data(action, space, signer, board, state, player1, player2):
+    if action == 'create':
+        if board is not None:
+            raise InvalidTransaction(
+                'Invalid action: Game already exists.')
+
+    elif action == 'take':
+        if board is None:
+            raise InvalidTransaction(
+                'Invalid action: Take requires an existing game.')
+
+        if state in ('P1-WIN', 'P2-WIN', 'TIE'):
+            raise InvalidTransaction(
+                'Invalid Action: Game has ended.')
+
+        if ((player1 and state == 'P1-NEXT' and player1 != signer)
+                or (player2 and state == 'P2-NEXT' and player2 != signer)):
+            raise InvalidTransaction(
+                "Not this player's turn: {}".format(signer[:6]))
+
+        if board[space - 1] != '-':
+            raise InvalidTransaction(
+                'Invalid Action: space {} already taken'.format(space))
+
+
+def _make_xo_address(name):
+    return '5b7349' + hashlib.sha512(name.encode('utf-8')).hexdigest()[:64]
```

Comment:
Passing in a namespace as an argument to the handler seems to me like an obsolete feature that can be removed, so I didn't want to get it involved. That said, if it's important, I can add it in to my changes.